core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/util/DatabaseUtil.java

```diff
@@ -598,13 +632,42 @@ private static void closeStatement(PreparedStatement preparedStatement) {
         if (preparedStatement != null) {
             try {
                 preparedStatement.close();
+            } catch (SQLRecoverableException ex) {
+                handleSQLRecoverableException(ex);
             } catch (SQLException e) {
                 log.error("Database error. Could not close statement. Continuing with others. - " + e.getMessage(), e);
             }
         }
 
     }
 
+    private static void handleSQLRecoverableException(SQLRecoverableException ex){
+		Connection dbConnection;
+    	try {
+	    	if(dataSource != null){
+	    		dbConnection = getDBConnection(dataSource);
+				handleSQLRecoverableException(dbConnection, ex);
+	    	}
+    	} catch (SQLException e) {
+            log.error("Error occurred during SQLRecoverableException handling." + e.getMessage());
+    	}
+    }
+    
+    	private static void handleSQLRecoverableException(Connection dbConnection, SQLRecoverableException ex) {
+        log.warn("Attempting recovery from thrown SQLRecoverableException ..." + ex.getLocalizedMessage(), ex);
+        try {
+        	if(dbConnection != null && !dbConnection.isClosed()){
+        			dbConnection.close();
+        		
+        	}else{
+            	String reason = dbConnection == null? "database connection is null" : dbConnection.isClosed()? "database connection is closed": "unknown";
+        		log.warn("Couldn't recover from SQLRecoverableException - cause: " + reason);
+        	}
+        } catch (SQLException e) {
+            log.error("Error occurred during SQLRecoverableException handling." + e.getMessage());
```

Comment:
It would be better to log the exception object without concatenating the message to the error log:
log.error("Error occurred during SQLRecoverableException handling.", e);